ui/FilesPanel.go

```diff
@@ -90,33 +90,58 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
 
 func (this *filesPanel) createBackButton() *gtk.Button {
 	image := utils.MustImageFromFileWithSize("back.svg", this.Scaled(40), this.Scaled(40))
-	return utils.MustButton(image, func() {
-		if this.locationHistory.Length() < 1 {
-			this.UI.GoToPreviousPanel()
-		} else if this.locationHistory.IsRoot() {
-			this.locationHistory.GoBack()
-			this.doLoadFiles()
-		} else {
-			this.locationHistory.GoBack()
-			this.doLoadFiles()
-		}
-	})
+	return utils.MustButton(image, this.goBack)
 }
 
 func (this *filesPanel) doLoadFiles() {
 	utils.EmptyTheContainer(&this.listBox.Container)
-
-	if this.displayRootLocations() {
-		this.addRootLocations()
-	} else {
+	if this.isRoot() {
+		if this.refreshSD() {
+			this.addRootLocations()
+		} else {
+			this.locationHistory = utils.LocationHistory {
+				Locations: []dataModels.Location{dataModels.Local},
+			}
+		}
+	}
+	if !this.isRoot() {
 		sortedFiles := this.getSortedFiles()
 		this.addSortedFiles(sortedFiles)
 	}
-
 	this.listBox.ShowAll()
 }
 
-func (this *filesPanel) displayRootLocations() bool {
+func (this *filesPanel) refreshSD() bool {
+	err := (&octoprintApis.SdRefreshRequest {}).Do(this.UI.Client)
+	if err == nil {
+		sdState, err := (&octoprintApis.SdStateRequest {}).Do(this.UI.Client)
+		if err == nil && sdState.IsReady == true {
+			return true
+		} else {
+			return false
+		}
+	} else {
+		return false
+	}
+}
+
+func (this *filesPanel) goBack() {
+	if this.isRoot() {
+		this.UI.GoToPreviousPanel()
+	} else if this.locationHistory.IsRoot() {
+		this.locationHistory.GoBack()
+		if this.refreshSD() {
+			this.doLoadFiles()
+		} else {
+			this.UI.GoToPreviousPanel()
+		}
+	} else {
+		this.locationHistory.GoBack()
+		this.doLoadFiles()
+	}
+}
+
+func (this *filesPanel) isRoot() bool {
```

Comment:
I like this function as `displayRootLocations` better than `isRoot`.  The reason is that in `goBack()`, there is:
```
func (this *filesPanel) goBack() {
	if this.isRoot() {
		this.UI.GoToPreviousPanel()
	} else if this.locationHistory.IsRoot() {
		this.locationHistory.GoBack()
```
We have `this.isRoot()` and `this.locationHistory.IsRoot()`.  When you look at the impl of isRoot, the impl is:
```
func (this *filesPanel) isRoot() bool {
	if this.locationHistory.Length() < 1 {
		return true
	} else {
		return false
	}
}
```
...and as a reader, questions come up, like
_"why define isRoot() when it just uses locationHistory.Length()?"_ 
and _"what is locationHistory.Length() vs locationHistory.IsRoot()?"_ 

When named `displayRootLocations()`, it's clearer as to what the intent of the logic is.

Reply:
PS. That being said, if you have a better name for `displayRootLocations`, I'm open to it being renamed.

Reply:
I'll go ahead and switch it back.  I am not sure about the name, personally, as it seems like it is an action, rather than a test, but best not to muck with it I guess.